mlflow/tracing/types/wrapper.py

```diff
@@ -1,32 +1,89 @@
 import json
 import logging
-from functools import lru_cache
-from typing import Any, Dict, Optional, Union
-
-from opentelemetry import trace as trace_api
-
-from mlflow.entities import Span, SpanContext, SpanEvent, SpanStatus, SpanType, TraceStatus
+from dataclasses import asdict
+from functools import lru_cache, wraps
+from typing import Any, Dict, List, Optional, Union
+
+from opentelemetry.sdk.trace import Event as OTelEvent
+from opentelemetry.sdk.trace import ReadableSpan as OTelReadableSpan
+from opentelemetry.trace import Span as OTelSpan
+
+from mlflow.entities.span import SpanType
+from mlflow.entities.span_event import SpanEvent
+from mlflow.entities.span_status import SpanStatus
+from mlflow.entities.trace_status import TraceStatus
+from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
 from mlflow.tracing.types.constant import SpanAttributeKey
-from mlflow.tracing.utils import TraceJSONEncoder, format_span_id, format_trace_id
+from mlflow.tracing.utils import (
+    TraceJSONEncoder,
+    build_otel_context,
+    decode_span_id,
+    decode_trace_id,
+    format_span_id,
+    format_trace_id,
+)
 
 _logger = logging.getLogger(__name__)
 
 
-class MlflowSpanWrapper:
+def live_span_only(func):
     """
-    A wrapper around OpenTelemetry's Span object to provide MLflow-specific functionality.
+    A decorator to ensure that the method is called only on "live"" spans. A live span is a span
```

Comment:
```suggestion
    A decorator to ensure that the method is called only on "live" spans. A live span is a span
```

Reply:
Is live equivalent to active?

Reply:
They are not same, and I know that's confusing but can't think of better word here... Basically, "live" is more broader state than "active". There is only one "active" span at the time that represents what the code is doing now. On the other hand, "live" just refers to the period since a span is started and until logged/saved.

For example, let's say we have two spans A and B created like this (the left digits are line numbers).
```
1.  with mlflow.start_span("A") as root_span:
2.     root_span.set_inputs("foo")
3.     with mlflow.start_span("B") as child_span:
4.         child_span.set_attirbutes("bar", 1)
5.         root_span.set_attributes("baz", 2)
6.     root_span.set_inputs("fuga")
7.   trace = mlflow.get_traces()[-1]
8.   span_A = trace.data.spans[0]
9.   span_B = trace.data.spans[1]
```

There can be only one **active** span at a time, so
- A is *active* at L1-2
- B is *active* at L3-5
- A becomes *active* again at L6-7.

On the other hand, multiple spans can be **live**.
- A is *live* at L1-L6.
- B is *live* at L2-L6.
The good example is L5 - at that moment, span B is active because it is the most deepest context manager. However, we can still set attribute to the span A because it is alive.

Reply:
Got it, thanks for the explanation!